Only allow users to specify a single location per parse call & pass full location data to schema.load

sirosen opened the pull request:
This is still incomplete, but core and flaskparser tests are now passing. No work is (yet) included on other concrete parsers, but I fixed the base asyncparser to make mypy linting pass.

Resolves #419, #267, #164, #268 
Obsoletes, and therefore closes, #410

I've updated the changelog with a basic "statement of intent". As I work on this, I will try to maintain it as an accurate record of how this change should look to the outside world. Obviously, narrative docs also need to be updated with relevant info.

Highlights / greatest hits:
- Marks this as an unreleased v6.0.0 in the changelog
- Replace `DEFAULT_LOCATIONS` with `DEFAULT_LOCATION="json"`
- Replace `parse_{location}` with `location_load_{location}`
- Replace `locations=<iterable>` with `location=<str>`
- Remove `locations=...` from Fields
- Remove `Parser.parse_arg`
- Remove `webargs.core.get_value`
- Add `webargs.multidictproxy.MultiDictProxy` which has `get_value`-like behavior and proxies various dict-like objects
- Reverse the decision made in #297 with [a comment in tests explaining this change](https://github.com/sirosen/webargs/blob/90c337278677aaca603b1b8f71aa914ceaad8dca/src/webargs/testing.py#L205-L213) and a bit of [similar explanation in the commit message](
https://github.com/marshmallow-code/webargs/commit/e59963f3994e26ee108226d52536af943ec9b961)
- Numerous fixes to tests to be more explicit about which location is being used (since you can't trivially have a single location `/echo` which loads data from a bunch of locations)
- As a side-effect of the above changes, certain inputs which were previously ignored are treated as errors (e.g. `1` for a JSON body will now parse and be passed literally to the schema, which will fail because `1` is not a valid input for `schema.load`). Changes to the tests serve as a somewhat-readable list of behavioral changes.

To try to keep this from sprawling *even more*, my intent is to address some items mentioned in #419 either in follow-up PRs or as part of this only if we're happy with this direction (and agree on those items). In particular: the `json_or_form` location I proposed and nested `use_args` usage and error collection.

@sloria, @lafrech, my biggest question is: do you like where this is going? Should I just hammer out all of the other parsers?

sirosen committed:
Update changelog with change in `location`

This is early documentation of planned changes, for this branch of work.

Note the change from `locations=...` to `location` in
Parser.use_[kw]args methods, and that fields may no longer specify a
location. Additionally, detail that schemas now load all data from a
location, not only those for which there are matching fields.

The changelog explains how to rewrite a schema which loads from multiple
locations into multiple `use_[kw]args` calls, and how to use `unknown`
to get webargs v5-like behavior.

sirosen committed:
Start to implement #419 (flaskparser first)

Stated goals in the form of CHANGELOG update

Gut significant chunks of webargs.core
- replace
    "parse_{location}" with "load_{location}"
    "locations=<collection>" with "location=<string>"
    "location_handler" with "location_loader"
- remove
    "parse_arg"
    "get_value"
- add
    webargs.multidictproxy which uses a Schema to handle MultiDict types
    and behave similarly to webargs.core.get_value

As a proving ground for this implementation/approach, get flaskparser
tests to pass under ma3

Fixes as the location loading code undergoes refactoring for the
flaskparser:
- Restore missing mimetype check for JSON payloads
- Treat form data as a multidict without extraneous AttributeError check
  (after testing, flask will return '{}' for form data when it is
  missing or empty -- just respect/allow this)
- req.headers is a multidict type as well
- fix some mistakes with new argument ordering

Changes to the base test app at src/webargs/testing.py of note:
- Because each route can only specify that it loads data from a single
  location (unless you pull extra tricks), several testing cases have
  had their routes modified, e.g. "/echo" -> "/echo_json"
- New test, test_parse_json_list_error_malformed_data added to check
  that webargs does not incorrectly treat non-list data as singleton
  lists when passing data to a schema with list fields. The previous
  behavior was questionable before, but with full data going to schemas
  it definitely becomes incorrect.
- Many tests which previously expected success with malformed data
  ignored now expect 422 errors. For example, sending a non-dict JSON
  value (e.g. `1`) should fail because it cannot be loaded by the
  schema. It should *not* be silently ignored because data was provided
  and failed to parse. This is a major behavior change for webargs.

Supporting modifications were made in tests/apps/flask_app.py ,
including the addition of `ma.EXCLUDE` when the detected marshmallow
version is >= 3.

After work on the relevant PR and trying things both ways, the
current handling of empty JSON bodies (interpreting them as {}) is being
kept externally, but changed internally. This relates to the decision
made in #297 . Originally in the course of work on #419, it seemed like
it might be necessary to change the behavior to make missing JSON data
behave as `missing` externally. However, making the change internal and
wrapping it to present the same external interface is best.

load_* methods are expected to return `missing` or {} when data is
missing for the given location. e.g. When a JSON body is "" it should be
`missing`. This is considered correct.
However, in terms of the external behavior of webargs, no data should be
interpreted as `{}`.

In order to achieve this, the wrapper which calls the various
"loader_func"s to load location data explicitly checks the result for
`missing` and converts it to `{}` if it is missing.

This makes it easy to use webargs to implement an API with optional JSON
data, but internally preserves the ability to detect a missing JSON
payload and distinguish it from a payload containing `{}`.

sirosen committed:
Update aiohttpparser for webargs v6

- Convert "parse_*" methods to "load_*" style
- Add needed methods to the testapp
- Update aiohttp tests in minor ways to handle changes
- Update asyncparser methods as well

sloria commented on CHANGELOG.rst:
```diff
@@ -13,6 +13,56 @@ Refactoring:
+        h1 = ma.fields.Int(location="headers")
+
+
+    @parser.use_args(CompoundSchema(), locations=("query", "headers"))
```
Maybe use dicts instead of schema classes for these examples; I think that's the more common usage.

sirosen replied:
Sure, I see now that other parts of the changelog use dicts. I'll make it consistent.

sloria commented on CHANGELOG.rst:
```diff
@@ -13,6 +13,56 @@ Refactoring:
 
+* *Backwards-incompatible*: Schema fields may not specify a location any
+  longer, and `Parser.use_args` and `Parser.use_kwargs` now accept `location`
+  (singular) instead of `locations` plural. Instead of using a single field or
```
```suggestion
  (singular) instead of `locations` (plural). Instead of using a single field or
```

sloria commented on src/webargs/asyncparser.py:
```diff
@@ -22,53 +21,12 @@ class AsyncParser(core.Parser):
     either coroutines or regular methods.
     """
 
-    async def _parse_request(
```
This will be nice cleanup. 👍

sirosen replied:
Aw, shucks. Thanks. 😊 

The refactor of the core parser is less obvious ("are we just shuffling around complexity?") but when you see how much this one changes and how much the diff is negative, it builds my confidence that this is really simplifying things.

sloria commented on src/webargs/flaskparser.py:
```diff
@@ -48,17 +49,20 @@ class FlaskParser(core.Parser):
+        return req.view_args or core.missing
 
-    def parse_json(self, req, name, field):
+    def location_load_json(self, req, schema):
```
How about `get_json` etc.? It's less verbose and no less clear, IMO.

sirosen replied:
Yeah, I'm fine with that. But I think I want to keep the name `_get_loader` for the wrapper which looks up these methods in the `__location_map__`. I can't think of a better name for that.

Another option is `load_json` if you like that. I'll rename to `get_{location}` throughout and I don't mind doing another rename if you want.

sirosen replied:
Actually, as I was in the midst of doing this, I saw that there are other functions like `get_default_request` and `get_request_from_view_args` which also have `get_` as a prefix. I'm going to go with `load_*` for now and we can discuss changing it. I'd like for the data-reading functions to have a distinct shared prefix that makes them identifiable.

sloria replied:
Fine with me

sloria commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+            if MARSHMALLOW_VERSION_INFO[0] < 3:
+                result.append(field.load_from or name)
+            else:
+                result.append(field.data_key or name)
```
`data_key` can be an empty string

```suggestion
                result.append(field.data_key if field.data_key is not None else name)
```

Same with `load_from`.

Just fixed this in marshmallow (https://github.com/marshmallow-code/marshmallow/pull/1379), so it's fresh in mind 😄

sirosen replied:
I see the truthiness of your words. (See what I did there? :wink: )
I think this is the only place now that inspects individual fields on the schemas, but I'll double check to make sure this isn't an issue elsewhere.

sloria commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+        self.multiple_keys = self._list_multiple_keys(schema)
+
+    def _list_multiple_keys(self, schema):
+        result = []
```
You can make this a `set` to optimize the membership checks.

sloria commented on src/webargs/testing.py:
```diff
@@ -186,19 +201,29 @@ def test_parse_files(self, testapp):
         )
-        assert res.status_code == 200
-        assert res.json == {"name": "World"}
+        assert res.status_code == 422
```
Agreed: this is probably the better behavior.

sloria commented on tests/test_core.py:
```diff
@@ -254,42 +243,37 @@ def test_handle_error_reraises_errors(web_request):
-    parse_json.side_effect = ValidationError("parse_json failed")
+    mock_schema = mock.Mock(spec=Schema)
+    mock_schema.strict = True
+    mock_schema.load.side_effect = ValidationError("parsing json failed")
```
Is the mocking here necessary?

sirosen replied:
This is a good catch. I'd rather use a real schema object where possible. Will fix.

sloria reviewed (commented):
Just drive-by's for now. Good work! Thanks for being so thorough with the docs.

I'll give this a proper lookover when I have more time to review.

sirosen commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+            if not is_multiple(field):
+                continue
+            if MARSHMALLOW_VERSION_INFO[0] < 3:
+                result.append(field.load_from or name)
```
```suggestion
                result.append(field.load_from if field.load_from is not None else name)
```

Adding a self-suggestion so that I can batch this with the GitHub suggestion commit for the change below.

sirosen committed:
Update bottleparser, djangoparser for v6

- Convert parse_* funcs to load_* funcs
- Update test apps to work with testsuite changes
- The django test app now decorates all view funcs so that validation
  errors get caught and translated to 422s (simplifies the test app)

To support these changes, add a "concrete" implementation of `load_json`
to the core parser.

Because the whole point of the explicit error handling is to try to make
the behavior of webargs more uniform, it makes sense to provide a couple
of hook points for sending data into the main load_json method.
The public interface is still the existence of load_json and users who
want to customize parser behavior should simply override that. However,
within webargs we can have some private hooks, `_raw_load_json` which
doesn't deal with decode errors and `_handle_invalid_json_error` which
processes errors.

As a result, bottleparser, flaskparser, and djangoparser can all share
code for this purpose. Presumably all of the other parsers will be able
to share as well, but it remains to be seen.

Per code review, update docstrings on all load_* methods returning
MultiDictProxy objects in core and flaskparser as well.

sloria commented on src/webargs/flaskparser.py:
```diff
@@ -72,31 +79,36 @@ def parse_json(self, req, name, field):
+        return MultiDictProxy(req.args, schema)
+
+    def load_form(self, req, schema):
+        """Read form values from the request.
```
Maybe simplify to

```suggestion
        """Return form values from the request as a MultiDictProxy."""
```

sirosen replied:
👍 I like this wording. I'll go through and update them all when I next get time to work on this.

sloria commented on tests/test_core.py:
```diff
@@ -59,66 +59,70 @@ def parser():
+)
+def test_parse_with_excluding_schema(parser, web_request):
+    """
+    This is new in webargs 6.x ; it's the way you can "get back" the behavior
```
Change this to use code comments instead of a docstring so that pytest doesn't output this text.

sloria commented on src/webargs/asyncparser.py:
```diff
@@ -77,17 +35,18 @@ class AsyncParser(core.Parser):
         Receives the same arguments as `webargs.core.Parser.parse`.
         """
-        self.clear_cache()  # in case someone used `parse_*()`
+        self.clear_cache()  # in case someone used `location_load_*()`
```
```suggestion
        self.clear_cache()  # in case someone used `load_*()`
```

sloria commented on src/webargs/core.py:
```diff
@@ -335,18 +238,18 @@ def parse(
 
          :return: A dictionary of parsed arguments
         """
-        self.clear_cache()  # in case someone used `parse_*()`
```
Why is this removed? I think it's still necessary.

sirosen replied:
That was a mistake which slipped through. I'll amend to revert this change (it's part of ab05a80c ).

In my very first drafts, I was playing with changing things about how the cache worked but it was just too much unrelated refactoring and noise. Apparently I accidentally added one of those changes.

sirosen replied:
Fixed this in recent changes ; thanks for spotting it.

(I'd rebase, but if you plan to squash, it's easier to review in new commits).

sloria commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+    def __delitem__(self, key):
+        del self.data[key]
+
+    def __setitem__(self, key, value):
```
Should we make these immutable?  Flask's req.form, req.args, etc return [ImmutableMultiDicts](https://github.com/pallets/werkzeug/blob/6817cd093ee642192604fff163fed1be0e0a191e/src/werkzeug/datastructures.py#L1579) anyway.

sirosen replied:
I don't think so, because it will fail with an appropriate error the way that this is now. Replicating the work in the frameworks to make these immutable just seems to offer little value to me.

If a Flask user ended up handling the `MultiDictProxy` (e.g. a validator which takes original data), using `__setitem__` will give a `TypeError` because it just calls `ImmutableMultiDict.__setitem__` anyway. That seems like pretty good behavior to me.

Is there a case in which this will fall over?

I am looking at those classes and wondering if we should proxy the `copy` method for Flask though. I wouldn't be surprised if other frameworks have useful specialty-methods which people might want to use too.

sloria replied:
Fair point; I think this is fine as is. I'm not sure about `copy`; I don't know when that would come up in practice.

sirosen replied:
Maybe the best option for `copy` and its ilk is to ignore it. And if someone comes along requesting that `MultiDictProxy` also proxies `copy`, it can be added pretty trivially.

sloria replied:
Sounds good

sloria commented on tests/apps/flask_app.py:
```diff
@@ -3,77 +3,100 @@
+if MARSHMALLOW_VERSION_INFO[0] < 3:
+    schema_kwargs = {"strict": True}
+else:
+    schema_kwargs = {"unknown": ma.EXCLUDE}
```
Is this necessary? Seems like the tests shouldn't be sending any unexpected keys.

sirosen replied:
There's currently a test which explicitly checks that an unexpected key is ignored -- behavior which shouldn't be the norm going forward. So for now this is just duct-tape to make that test work.

I think we want tests, probably with `skipif`s for marshmallow 2, which explores the various `unknown` settings you could pass to the schema, to replace this.

I'll make that change a priority, so that we can talk about whether or not we like those tests.

sirosen replied:
05d308e changes this around a bit.
The `unknown=EXCLUDE` is now used (1) in an explicit "ignores extra" route and (2) in `/echo_headers`, to handle the presence of various headers not explicitly set by the client.

sloria commented on tests/test_flaskparser.py:
```diff
@@ -33,23 +33,31 @@ def test_use_args_with_view_args_parsing(self, testapp):
+            "/echo_nested_many_data_key",
+            {"x_field": [{"id": 42}]},
+        )
+        # under marhsmallow2 this is allowed and works
```
```suggestion
        # under marshmallow 2 this is allowed and works
```

sirosen replied:
This was done in 176a9f7 (and I marked co-authorship the way github does out of curiosity, to see if it would understand and resolve this suggestion).
It doesn't seem to have, so I'm just going to mark the accepted suggestions as resolved.

sloria reviewed (commented):
Reviewed this a bit more. I'm basically good with the approach here.

I cut 5.5.0, so you'll need to update with `dev`. Feel free to merge instead of rebase; I intend to squash and merge when this is all done, if you don't mind.

I also made a 5.x-line branch so that we can maintain support for 5.x for some time after we release 6.0.

sloria commented:
Should we rename `@location_handler` to `@location_loader` for consistency with the `load_*` naming?

Just a thought: doesn't need to happen in this PR.

sirosen committed:
Make tests explicit about ignoring extra data

Some routes/methods are being tested to ensure that they accept
additional data. In those cases and *only those cases*, the
unknown=EXCLUDE argument should be passed to the schema when running
under marshmallow 3.

This schema must be used for handling headers as well, as otherwise
normally seen headers like `Host` will cause schema validation errors.

Additionally, fix an issue with the flaskparser under marshmallow2: the
validation of a "many=True" schema against non-list data produced error
messages in a different format from marshmallow 3 (use of `enumerate()`
expanding out in a presumably-unexpected way). In order to compensate,
add a check under marshmallow 2 for precisely this condition and
reformat the error message in this case.
As a result of this fix, tests are now passing for the flaskparser under
all tox environs.

Updates to bottleparser and djangoparser tests to match.

This also reverts some changes to the test apps which turned out to be
unnecessary.

sirosen commented:
> Should we rename `@location_handler` to `@location_loader` for consistency with the `load_*` naming?

I don't mind changing it or keeping it the same. There isn't a huge difference between these names.

One advantage of renaming is that it would immediately fail for webargs v5 users who are using `@location_handler` today and upgrade. That might be reason enough to make the change.

sloria commented:
Up to you. Feel free to do the `location_loader` rename, or we can do it in a follow-up PR.

sirosen committed:
Convert remaining parsers for v6

Updates falconparser, webapp2parser, tornadoparser, and pyramidparser.

For the most part, this was a matter of
- Convert parse_* to load_* methods, using MultiDictProxy
- Update tests and test apps

There are pecularities for Falcon and Tornado.

- Some notes inline in falconparser about use of MultiDictProxy and
  handling of headers and cookies, as Falcon does not provide a
  high-level multidict representation of these things. (Webargs
  *could* provide one with great effort if it's ever wanted.)

- In the case of Tornado, some custom multidict classes are needed to
  customize the handling of data.

sirosen commented:
> Up to you. Feel free to do the `location_loader` rename, or we can do it in a follow-up PR.

I'll probably do it in here, since it's already a large, breaky change.

---

Something that I've started to realize poses an issue: the way that I've been writing things, such that webargs treats an empty JSON payload as `missing`, actually makes the behavior more confusing and non-uniform.

The issue is that `json` is the *only* location which can behave this way. No query params or an empty form post needs to be interpreted as `{}`.

I'm working on a refactor now in which `load_json` will still return `missing`, but `_load_location_data` (the wrapper which gets the loader func and calls it) will check if the result is `missing` and return `{}` if it is.
This lets webargs can still pass around information about the presence or absence of data, but when data is being handed off to `Schema.load` it needs to be "normalized" for consumption.

Part of the goal is to ensure it's easy to add something like
```python
def load_json_or_form(self, req, schema):
    data = self.load_json(req, schema)
    if data is missing:
        data = self.load_form(req, schema)
    return data
```

and I think this achieves that.

This will have a public-facing impact on how webargs v6 behaves, but I think generally for the better. It will make it easier to use webargs in APIs with optional request bodies.
So expect it to come with numerous changes to tests' expectations around posting empty data to `/echo_json` endpoints.

sirosen commented:
Er, to clarify, I think the above is important to be very clear and open about because it constitutes another reversal on #297 (back to the current behavior)

The first version of this PR changed the decision on that saying that `""` is invalid JSON.
I thought this change would be necessary for implementing `json_or_form`, and it kind of appeals as more pedantically correct. But as I work on things I realize that this has negative impact on webargs' usability, and it's not as necessary as I thought if we distinguish between the value returned by `load_json` and the value seen by `Schema.load`.

One of the big cases that changes my mind is the idea of an endpoint with an optional JSON body.
There are a number of behaviors currently accepted for a missing request body, and it's not really obvious what should be allowed on this front.

sirosen commented:
As of f005c40 , this is finally be passing tests, as all of the parser conversions are finished. And I've just merged `dev` back in again to get this up to date.

This is what I'd consider the very minimum to be safe to merge. But I still plan to do two more things before removing `WIP` from the PR title:
- Update narrative docs with new usage
- Add some explicit testing of various `unknown=...` behaviors

sloria commented on src/webargs/aiohttpparser.py:
```diff
@@ -147,7 +148,7 @@ def handle_error(
         schema: Schema,
         error_status_code: typing.Union[int, None] = None,
         error_headers: typing.Union[typing.Mapping[str, str], None] = None,
-    ) -> "typing.NoReturn":
```
This is actually necessary to account for older patch versions of Python 3.5 https://github.com/marshmallow-code/webargs/issues/366

sirosen replied:
That makes me sad, but makes sense. I'll switch it back.

sloria replied:
Yeah.. I'm getting kinda exhausted by [these](https://github.com/marshmallow-code/marshmallow/pull/1410) =/

sloria replied:
Alternatively, we could conditionally install `typing` on Python 3.5. Something like:

```python
'typing>=3.7.4; python_version<"3.5.4"',
```

I think I'm going to end up doing this in marshmallow. 

Doesn't have to happen in this PR though.

sloria replied:
Disregard that; adding the typing backport doesn't fix it. It's only useful for Python versions less than 3.5.0.

sloria reviewed (commented):
Excellent. Let me know when you're confident in merging this.

sirosen committed:
Rewrite narrative docs to be correct in v6

This is not a deep and comprehensive rewrite which aims to
discuss/explain the functionality which is now available. Rather, this
change merely takes everything in the docs which has become inaccurate
and trims or modifies it to be accurate again.

Add notes to changelog about making parsers more consistent about
checking Content-Type for JSON payloads.

sirosen committed:
Elaborate on core test for unknown=... behavior

Rather than just comparing default schema behavior against
unknown=EXCLUDE, compare IGNORE and RAISE as well.

sirosen committed:
Add the 'json_or_form' location + docs

'json_or_form' is defined as first trying to load the JSON data, then
falling back to form data.

Adds a test to the testsuite for this which tries sending JSON, tries
sending Form, and tries sending no data, verifying that we get the
correct result in each case.

This does *not* try to support more elaborate behaviors, e.g. multipart
bodies, which probably will work variably well or poorly based on the
framework in use. In these cases, the content type will not detect as
JSON (because it is 'multipart'), and we'll failover to using form data.
This is acceptable, since anyone using `json_or_form` is expecting to
allow form posts anyway.

Docs cover not only the new json_or_form location, but also the *idea*
of a "meta location" which combines multiple locations. This is worth
documenting in the advanced docs since it is the way to get back certain
behaviors from webargs v5.x

sirosen committed:
Ensure all parsers treat missing json equivalently

If an empty body is submitted to a webargs parser, it should always be
checked for content-type. Otherwise, the behavior framework-to-framework
for webargs will not be consistent.

Namely: djangoparser and pyramidparser now check content-type for JSON
and return missing when an emtpy/missing payload is used. This makes the
detection/use of JSON payloads more consistent and is necessary for
`json_or_form`.

sirosen commented:
I've just removed `[WIP]` from the title because I think this is ready.
I've made changes to the narrative docs (more content was removed than added; I am sorry), and added some more tests.

As an important final change, I was trying to add `json_or_form` as a location -- to validate that the approach I have for that will work -- and my testing revealed some non-uniform treatment of empty JSON payloads between different frameworks.

I've added https://github.com/marshmallow-code/webargs/pull/420/commits/b8d3fee2c9d332c3ca3f3ba49ea0c483da18a590 , which makes the decision to check content-type on the frameworks where this was not done (django and pyramid) and return `missing` if it is non-JSON.

lafrech commented on docs/quickstart.rst:
```diff
@@ -243,7 +237,7 @@ Nesting Fields
 .. note::
 
-    By default, webargs only parses nested fields using the ``json`` request location. You can, however, :ref:`implement your own parser <custom-parsers>` to add nested field functionality to the other locations.
+    Of the default supported locations in webargs, only the ``json`` request location supports nested datastrctures. You can, however, :ref:`implement your own data loader <custom-loaders>` to add nested field functionality to the other locations.
```
s/datastrctures/datastructures/

sirosen replied:
Good catch. Let me patch this right now so I don't forget.

sloria commented:
Great. I probably won't be able to review this until this weekend at the earliest. Thanks for all your work on this.

sirosen commented:
Just as a short FYI: I'm going to be taking a couple of weeks off, and likely won't be able to respond to any feedback for the next two weeks or so anyway. So please don't worry if you write some comments and don't see a reply -- this isn't abandoned, and I'm still invested in seeing it through.

sirosen commented:
In light of the fact that it motivates b8d3fee , I've [added the `json_or_form` location to this branch](https://github.com/marshmallow-code/webargs/pull/420/commits/1d352d2ad49e1eb503db97e32f88f73eebe16de7). The former commit doesn't make as much sense without the latter.

lafrech commented:
Thank you so much @sirosen for tackling this and sorry for the so long delay.

I've tried to port flask-smorest to your webargs branch and I only had two adaptations to make:

- replace `locations=[location]` with `location=location` in a call to use_args
- add `location='query'` to a call to `parser.parse`

Those are known breaking changes. My point is that I didn't discover any issue.

Then I tried to port the application we're working on. I had to

- fix `NestedQueryArgsParser`. We copied it from webargs docs and since you also fixed the docs we just had to copy your new version. Great.
- add missing `content_type='application/json'` in some test functions

Again, nothing surprising.

I shall now actually review the code, but at least it proved functional on our real-life cases.

Thanks again.

sirosen commented:
I'll take some time soon (hopefully next few days) to revisit the changelog and make sure it's accurate. In particular

> add missing `content_type='application/json'` in some test functions

made me realize that two parsers have changed (the `DjangoParser` and `PyramidParser`) to require `Content-Type` where they did not before. That's currently missing from the changelog.

---

Anyway, please don't worry about taking time with this! I _more than_ understand how hard it can be to get time to work on open source projects and properly review big PRs. 🙂

lafrech commented:
Thanks.

I'm using Flask, BTW, so I guess this also changed on  `FlaskParser`.

lafrech commented:
I could be wrong but here's my understanding:

- In the Flask client, if I forget `content_type='application/json'`, the payload is sent as form data.
- webargs used to check several locations by default and now it only checks `json`.

So the breaking change would only be the default location checked for input.

sirosen commented:
I just amended the changelog for 6.0 based on a reread of the changes. Listed as a bugfix (because it makes webargs internally consistent), several parsers now require `Content-Type` to be set for JSON payloads. Since anything matching `application/(.*\+)?json` will do, I didn't specify exactly what it must be set to.

`FlaskParser` was one of the parsers with changes. Even if you send it JSON data, if you don't sent the `Content-Type` header it will not be parsed as JSON.
That jives with what you're saying you observed, so I think all is well.

---

There are new changes in `dev`, but they didn't merge back into this branch smoothly (quickly "fixing" the conflicts produced broken tests). I'll have to look more closely at the changes to understand what went wrong.

sirosen commented:
I've just rebased and pushed a cleaned-up version of this branch which works against `dev`. (I tried doing this merge-back thing, but it's just not how I'm used to using git. The commit history becomes harder to navigate.)

My attempt to handle this by merging things yesterday failed because I did not properly incorporate #428 . Because this PR introduces the idea of a shared `load_json` handler in the core parser, that needed to be updated as well.

lafrech commented:
Thanks.

Rebase is better. Sometimes it is harder to achieve (too many changes and too many commits in the branch to rebase), so a merge is acceptable. If rebase is easier for you, then that's just great.

I'd really like to merge this. I'll try to review this as soon as I can. Can't promise when.

lafrech commented on docs/quickstart.rst:
```diff
@@ -23,17 +23,11 @@ Arguments are specified as a dictionary of name -> :class:`Field <marshmallow.fi
-        "profile_image": fields.Field(
-            location="files", validate=lambda f: f.mimetype in ["image/jpeg", "image/png"]
-        ),
+        # "content_type": fields.Str(data_key="user-type"),
```
s/content_type/user_type/ ?

sirosen replied:
I need to take another look at this, but obviously there's a typo here.

I think really that I'd like to revisit the quickstart after working on stacked `use_args` errors ([per this comment from sloria](https://github.com/marshmallow-code/webargs/issues/419#issuecomment-525765453)), since that will make it possible to have use-cases in here with
```python
@use_args(...)
@use_args(...)
def view_func(...):
    ...
```

(Those work today, but error semantics are based on the order of the decorators.)

sirosen replied:
I reread the doc. It turns out, fixing to just read `user_type`, as you suggested, is simplest.

lafrech commented on src/webargs/core.py:
```diff
@@ -531,41 +446,98 @@ def handle_error(error, req, schema, status_code, headers):
+                return self._handle_invalid_json_error(e, req)
+
+        return json_data
+
```
Why not also cache `missing`?

```py
        json_data = self._cache.get("json")
        if json_data is None:
            try:
                json_data = self._raw_load_json(req)
            except json.JSONDecodeError as e:
                if e.doc == "":
                    json_data = missing
                else:
                    return self._handle_invalid_json_error(e, req)
            except UnicodeDecodeError as e:
                return self._handle_invalid_json_error(e, req)

        self._cache["json"] = json_data
        return json_data
```

sirosen replied:
I missed this in my pass over things today.
If I had to guess, I'd say I was just imitating what came before. I see no reason not to cache `missing` results -- I'll make the change.

sirosen replied:
I've just changed this. One test which explicitly checked `self._cache` needed to be tweaked, but nothing else broke with the change.
It's simpler, but I noted in the commit message that we're changing it in part because we don't have context here for why we _shouldn't_ cache `missing`. So if it turns out, on review, that there's some issue which comes up in webargs 6.x because of it, we won't be worried about reverting the change.

lafrech commented on src/webargs/core.py:
```diff
@@ -329,9 +239,9 @@ def parse(
+        :param str location: Where on the request to load values.
+            Can be one of ``('json', 'querystring', 'form', 'headers', 'cookies',
+            'files')``.
         :param callable validate: Validation function or list of validation functions
```
The list is incomplete. Should we point to `__location_map__` instead?

sirosen replied:
I'm going to take a look at this, but my instinct is to both write it all out here and point there in addition... In terms of the documentation's readability, I think it helps to have the list of available values included right there.

sirosen replied:
I've made a change to do both. I was a little concerned that there might be other locations that need to be kept in sync, but it looks like this is the only place where the contents of `__location_map__` are spelled out.

I think this is an okay solution, so I'm going to mark this resolved. But I don't mind discussing further if you like -- e.g. if you're worried that it will get out of sync again in the future.

lafrech commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+            else:
+                return [val]
+        return val
+
```
no-else-return: Unncessary "else" after "return"

sirosen replied:
I'm not sure I follow. You mean it should read
```python
if val is None:
    return None
return [val]
```
?

I don't mind doing that if you think it reads cleaner to drop the `else` (or if its a general style rule for certain style guides).

lafrech replied:
This is something that pylint warns about. Removing the else after every return makes it clearer that the return line exits the function. At least it is the rationale.

```py
    def __getitem__(self, key):
        val = self.data.get(key, missing)
        if val is not missing and key in self.multiple_keys:
            if hasattr(self.data, "getlist"):
                return self.data.getlist(key)
            if hasattr(self.data, "getall"):
                return self.data.getall(key)
            if isinstance(val, (list, tuple)):
                return val
            if val is None:
                return None
            return [val]
        return val
```

or even

```py
    def __getitem__(self, key):
        val = self.data.get(key, missing)
        if val is missing or key not in self.multiple_keys:
            return val
        if hasattr(self.data, "getlist"):
            return self.data.getlist(key)
        if hasattr(self.data, "getall"):
            return self.data.getall(key)
        if isinstance(val, (list, tuple)):
            return val
        if val is None:
            return None
        return [val]
```

sirosen replied:
I like the second version better because it's less nested, so I switched to that.

I tried to take a look at what pylint has to say to see if there's anything else I can trivially clean up, but I pretty much instantly re-learned why I find pylint so annoying. 😂 
(It's a good tool and all, but it is _picky_!)

lafrech commented on src/webargs/multidictproxy.py:
```diff
@@ -0,0 +1,67 @@
+    Whenever a value is looked up, it is checked against the schema to see if
+    there is a matching field where `is_multiple` is True. If there is, then
+    the data should be loaded as a list or tuple.
+
```
I would have been tempted to subclass rather than proxying, to only override what is needed, but I suppose there's a good reason for proxying.

sirosen replied:
It's hard to be sure exactly what I was thinking when I wrote the first version, but I think that proxying ends up being less code. If you subclass the multidict types for Flask, Django, etc. you end up with a variety of classes to handle.

Also, if I'm right that Falcon really does need MultiDictProxy for List fields (I need to take some time to prove this out), then we basically end up writing this up anyway, since Falcon doesn't have any native multidict type.

lafrech commented on src/webargs/falconparser.py:
```diff
@@ -91,12 +75,20 @@ def to_dict(self, *args, **kwargs):
+    # MultiDictProxy because we need to use the schema to determine when single
+    # values should be wrapped in lists due to the type of the destination
+    # field
+
```
I wasn't sure about this so I removed the `MultiDictPRoxy`. No test failed.

I didn't investigate any further but maybe a test could be added.

sirosen replied:
It's hard to remember perfectly, but based on the comment I think this is something to do with the interplay between `marshmallow` Lists and Falcon parsing.
I need to go in and add a test to try to capture this.

sirosen replied:
I confirmed that there's a failing case if you remove this and I've added a test for it.
If you remove `MultiDictProxy` and pass a _single value_ to a `List` field, then things will fail.

The reason is that Falcon parses query params like so:
```
/path?foo=1&foo=2
  =>  {"foo": [1, 2]}
/path?foo=1
  =>  {"foo": 1}
```
When we try to convert `{"foo": 1}` to a list field, we run aground on `1` as a non-list value.

lafrech reviewed (commented):
Looks good to me.

I made a few comments.

I didn't review the async parsers.

I lean towards releasing as 6.0 rather than as a 6.0b pre-release.

sirosen reviewed (commented):
I wanted to get to the review feedback this past weekend, but it didn't happen. I'll try to find a good time for it soon. There's one item where I had a question to make sure I correctly understand what's wanted.

sirosen committed:
Improvements per @lafrech's review

1. Fix a typo in the quickstart doc

Rewrite "content_type" to "user_type". Just a slip-up.

2. Flatten out the nesting in MultiDictProxy.__getitem__

Although this was flagged as an `else-return` which could flatten
out to just a return, Jérôme also suggested a version which treats the
outer condition as an `if ...: return` so that we can dedent the rest
of the logic.
The end-result is easier to read, so let's go with that.

3. Add a test for passing a single value to a List field

This applies to all parsers, but it was required by FalconParser in
particular. If you remove MultiDictProxy from the presentation of
query params, headers, etc. from FalconParser, it works in most cases.
However, because field values only get wrapped in lists by that parser
if there are multiple values supplied, we need some intermediary to
understand that lookups for List fields always need a `list`-type
value. MultiDictProxy serves that purpose, as a schema-aware dict type.

sirosen reviewed (commented):
I just pushed a commit with fixes for the several issues found in the last review.
I'm going to do a final double-check on everything and (hopefully) mark all of the outstanding comments as resolved.

sirosen committed:
Switch JSON parsing to cache `missing` values

Previously, `load_json` carefully avoided putting `missing` into the
`Parser._cache["json"]` storage location. However, we're not clear (at
least, in the context of #420 ) on why we should avoid caching
`missing` values. No tests fail, so change this to be simpler logic in
which any "successful" JSON parsing, including those which failover to
`missing`, will be cached.

sirosen committed:
Fix Parser.parse docstring for 'location'

Refer to the `__location_map__` attribute, and expand the list of
supported locations to be complete (add `query` and `json_or_form`).

sirosen commented:
Okay, I missed a couple of small things, so I handled those in follow-up commits.

I'm excited about merging this in part because I want to get cracking on the changes to `@use_args` I pitched in #419 for us to review. 😄

lafrech commented:
Great.

@sloria, should we release a beta to give frameworks time to adapt to the changes?

Then we'd release a final 6.0.0 with [Python 2 dropped](#440).

sloria commented:
Haven't had time to review the latest changes, so I'll defer to @lafrech on merging this. Releasing a 6.0.0b1 makes sense to me

lafrech commented:
Alright. Merging. I'll try to release a beta version today.

lafrech commented:
@sloria, I just got this error again.

```
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/dev.
remote: error: Required status check "marshmallow-code.webargs" is expected.
To ssh://github.com/marshmallow-code/webargs
 ! [remote rejected] dev -> dev (protected branch hook declined)
error: failed to push some refs to 'ssh://git@github.com/marshmallow-code/webargs'
```

Am I missing privileges on this repo?

sloria commented:
@lafrech I just gave you admin privs. Can you try again?
